### PR TITLE
TdS prize tool

### LINKED ISSUE #6077: TdS Compensation: Compute rewards earned for each validator
Related to: https://github.com/solana-labs/solana/issues/6076

#### Problem
The TdS Compensation tool needs to find the top rewards earners in the TdS ledger

#### Proposed Solution
Rewards should be calculated as the sum of all vote and stake account balances attributed to a particular validator pubkey minus the initial allotment of tokens at the beginning of stage. In order to calculate this sum, the `ledger-tool` will need to be updated to read ending account balances and state.

Note that validator account balances will **not** be counted, only vote and stake balances will be considered.

In general, stakers are expected to redeem their credits as they are earned so it should be reasonable to rely on that behavior to compute rewards in this way.

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Need transparent deterministic tooling for determining the winners of TdS

#### Summary of Changes
1. Create a new CLI tool which behaves similarly to `ledger-tool` for the purpose of computing the TdS prize winners. This new tool will be moved to https://github.com/solana-labs/tour-de-sol once required runtime changes are shipped to crates.io
2. Compute the winners for the "Most Rewards Earned" category ([TdS Compensation Design](https://forums.solana.com/t/tour-de-sol-stage-1-preliminary-compensation-design/79)) by summing the balances of the vote and stake accounts which are attributed to a particular validator.

Fixes #6077
Depends on #6184

### CODE REVIEW

File: tds-prize-tool/src/rewards_earned.rs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+fn voter_stake_rewards(stake_accounts: HashMap<Pubkey, Account>) -> HashMap<Pubkey, u64> {
+    let mut voter_stake_sum: HashMap<Pubkey, u64> = HashMap::new();
+    for (_key, account) in stake_accounts {
+        if let Some(StakeState::Stake(_authorized, _lockup, stake)) = StakeState::from(&account) {
+            let stake_sum = voter_stake_sum
+                .get(&stake.voter_pubkey)
+                .cloned()
+                .unwrap_or_default();
+            voter_stake_sum.insert(stake.voter_pubkey, stake_sum + account.lamports);
```
Comment (rob-solana) [inline tds-prize-tool/src/rewards_earned.rs]: strictly speaking the staked amount is a function of epoch and stake_history, this simplification can only be made if we ignore warmup

File: tds-prize-tool/src/rewards_earned.rs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    let mut validator_rewards: Vec<(Pubkey, u64)> = validator_reward_map
+        .iter()
+        .map(|(key, balance)| (*key, *balance))
+        .collect();
+
+    // Sort descending and calculate results
+    validator_rewards.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
+    validator_rewards
+        .into_iter()
+        .map(|(key, earned)| (key, (earned as i64) - STARTING_BALANCE_LAMPORTS))
```
Comment (rob-solana) [inline tds-prize-tool/src/rewards_earned.rs]: might want to take STARTING_BALANCE_LAMPORTS as a parameter for this tool?

File: tds-prize-tool/src/rewards_earned.rs (hunk 1 above)
Comment (jstarry) [reply tds-prize-tool/src/rewards_earned.rs]: Yeah, this tool isn't calculating total stake. It's calculating the total of lamports attributed to a validator through the balances of stake accounts and vote accounts according to @mvines suggestion here https://github.com/solana-labs/solana/issues/6077#issuecomment-535564165

File: tds-prize-tool/src/rewards_earned.rs (hunk 2 above)
Comment (jstarry) [reply tds-prize-tool/src/rewards_earned.rs]: @mvines are all validators starting with the same balance?

File: tds-prize-tool/src/rewards_earned.rs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    vote_accounts: HashMap<Pubkey, (u64, Account)>,
+) -> HashMap<Pubkey, u64> {
+    // Sum validator earned reward totals (stake rewards + commission)
+    let mut validator_reward_map: HashMap<Pubkey, u64> = HashMap::new();
+    for (voter_key, (_stake, account)) in vote_accounts {
+        if let Some(vote_state) = VoteState::from(&account) {
+            let voter_commission = account.lamports;
+            let voter_stake_reward = voter_stake_rewards.remove(&voter_key).unwrap_or_default();
+
+            // TODO: check if it's possible to have multiple vote accounts in working bank
```
Comment (rob-solana) [inline tds-prize-tool/src/rewards_earned.rs]: it is ;)

Comment (jstarry) [reply tds-prize-tool/src/rewards_earned.rs]: 🙊

Comment (rob-solana) [reply tds-prize-tool/src/rewards_earned.rs]: it's not a reasonable setup, as our implementation only supports voting to one account at a time, but it's possible to have stake dedicated to a single node via 2 accounts

File: tds-prize-tool/src/rewards_earned.rs (hunk 2 above)
Comment (mvines) [reply tds-prize-tool/src/rewards_earned.rs]: yes, in that everybody starts with a system account with 1000 SOL in it.

**Comment by mvines** [comment]
> This new tool will be moved to https://github.com/solana-labs/tour-de-sol once required runtime changes are shipped to crates.io

Is it practical to cherry-pick what you need into the 0.19 branch?  We're going to cut a 0.19.1 release within a day or so I expect, so if that ships then we can just keep this tool directly in the tds repo

File: tds-prize-tool/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    let blocktree = match Blocktree::open(&ledger_path) {
+        Ok(blocktree) => blocktree,
+        Err(err) => {
+            eprintln!("Failed to open ledger at {:?}: {}", ledger_path, err);
+            exit(1);
+        }
+    };
+
+    println!("Verifying ledger...");
+    match process_blocktree(&genesis_block, &blocktree, None, true, None) {
```
Comment (mvines) [inline tds-prize-tool/src/main.rs]: Def want `verify_ledger = false`, verifying PoH in this context probably borders on paranoia

File: tds-prize-tool/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,3 @@
+pub(crate) fn sol_to_lamports(sol: f64) -> u64 {
```
Comment (mvines) [inline tds-prize-tool/src/utils.rs]: We should sneak this into sdk/ somewhere.

**Comment by jstarry** [comment]
> Is it practical to cherry-pick what you need into the 0.19 branch?

Good call, here's the cherry pick: https://github.com/solana-labs/solana/pull/6190

> We're going to cut a 0.19.1 release within a day or so I expect, so if that ships then we can just keep this tool directly in the tds repo

Sounds good, I'll move this over there then and depend on the v0.19 github branch until 0.19.1 is released

File: tds-prize-tool/src/utils.rs (hunk 5 above)
Comment (jstarry) [reply tds-prize-tool/src/utils.rs]: Sounds good: https://github.com/solana-labs/solana/pull/6192

**Comment by jstarry** [comment]
Closed in favor of https://github.com/solana-labs/tour-de-sol/pull/12

**Pull request closed without merging**